django/forms/widgets.py

```diff
@@ -115,39 +117,54 @@ def __getitem__(self, name):
         raise KeyError('Unknown media type "%s"' % name)
 
     @staticmethod
-    def merge(list_1, list_2):
+    def merge(*lists):
         """
-        Merge two lists while trying to keep the relative order of the elements.
-        Warn if the lists have the same two elements in a different relative
-        order.
+        Merge several lists while trying to keep the relative order of the elements.
+        Warn if the lists have the same elements in a different relative order.
 
         For static assets it can be important to have them included in the DOM
         in a certain order. In JavaScript you may not be able to reference a
         global or in CSS you might want to override a style.
         """
-        # Start with a copy of list_1.
-        combined_list = list(list_1)
-        last_insert_index = len(list_1)
-        # Walk list_2 in reverse, inserting each element into combined_list if
-        # it doesn't already exist.
-        for path in reversed(list_2):
-            try:
-                # Does path already exist in the list?
-                index = combined_list.index(path)
-            except ValueError:
-                # Add path to combined_list since it doesn't exist.
-                combined_list.insert(last_insert_index, path)
-            else:
-                if index > last_insert_index:
-                    warnings.warn(
-                        'Detected duplicate Media files in an opposite order:\n'
-                        '%s\n%s' % (combined_list[last_insert_index], combined_list[index]),
-                        MediaOrderConflictWarning,
-                    )
-                # path already exists in the list. Update last_insert_index so
-                # that the following elements are inserted in front of this one.
-                last_insert_index = index
-        return combined_list
+
+        dependency_graph = defaultdict(set)
+
+        # Annotate items with the order in which they were seen making
+        # the resulting list order deterministic.
+        counter = count()
+        order = defaultdict(lambda: next(counter))
+
+        for lst in lists:
+            if not lst:
+                continue
+            base = (order[lst[0]], lst[0])
+            # First entries are always part of the dependency graph.
+            dependency_graph.setdefault(base, set())
+            for item in lst[1:]:
+                next_ = (order[item], item)
+                # No self dependencies
+                if base != next_:
+                    dependency_graph[next_].add(base)
+                base = next_
+
+        # Sort sublists by the counter value assigned above and discard
+        # the value afterwards.
+        try:
+            return list(chain.from_iterable(
```

Comment:
You might want to consider returning a generator yourself. In which case you should just `yield from` the generator instead of returning a list.

Reply:
Thanks. I noticed that I should use the `stable_topological_sort` function instead of creating my own logic with `order` dicts and counters. `stable_topological_sort` always returns a list which we can return without any additional modifications.